### PR TITLE
[SPARK-14680][SQL]Support all datatypes to use VectorizedHashmap in TungstenAggregate

### PR DESCRIPTION
**Pull request by sameeragarwal**
## What changes were proposed in this pull request?

This PR adds support for all primitive datatypes, decimal types and stringtypes in the VectorizedHashmap during aggregation.
## How was this patch tested?

Existing tests for group-by aggregates should already test for all these datatypes. Additionally, manually inspected the generated code for all supported datatypes (details below).

### CODE REVIEW

**Comment by sameeragarwal** [comment]
test this please

**Comment by sameeragarwal** [comment]
test this please

**Comment by sameeragarwal** [comment]
test this please

**Comment by sameeragarwal** [comment]
Generated VectorizedHashMap class for a contrived query that groups by against a range of datatypes: 

```scala
sqlContext.range(N)
  .selectExpr(
    "id" ,
    "(id & 65535) as k0",
    "cast(id & 65535 as string) as k1",
    "cast(id & 65535 as int) as k2",
    "cast(id & 65535 as double) as k3",
    "cast(id & 65535 as float) as k4",
    "id > 100 as k5")
  .groupBy("k1", "k2", "k3", "k4", "k5")
  .sum()
  .collect()
```

looks like:

```java
/* 075 */   public class agg_VectorizedHashMap {
/* 076 */     private org.apache.spark.sql.execution.vectorized.ColumnarBatch batch;
/* 077 */     private org.apache.spark.sql.execution.vectorized.ColumnarBatch aggregateBufferBatch;
/* 078 */     private int[] buckets;
/* 079 */     private int capacity = 1 << 16;
/* 080 */     private double loadFactor = 0.5;
/* 081 */     private int numBuckets = (int) (capacity / loadFactor);
/* 082 */     private int maxSteps = 2;
/* 083 */     private int numRows = 0;
/* 084 */     private org.apache.spark.sql.types.StructType schema =
/* 085 */     new org.apache.spark.sql.types.StructType()
/* 086 */     .add("k1", org.apache.spark.sql.types.DataTypes.StringType)
/* 087 */     .add("k2", org.apache.spark.sql.types.DataTypes.IntegerType)
/* 088 */     .add("k3", org.apache.spark.sql.types.DataTypes.DoubleType)
/* 089 */     .add("k4", org.apache.spark.sql.types.DataTypes.FloatType)
/* 090 */     .add("k5", org.apache.spark.sql.types.DataTypes.BooleanType)
/* 091 */     .add("sum", org.apache.spark.sql.types.DataTypes.LongType)
/* 092 */     .add("sum", org.apache.spark.sql.types.DataTypes.LongType)
/* 093 */     .add("sum", org.apache.spark.sql.types.DataTypes.LongType)
/* 094 */     .add("sum", org.apache.spark.sql.types.DataTypes.DoubleType)
/* 095 */     .add("sum", org.apache.spark.sql.types.DataTypes.DoubleType);
/* 096 */     
/* 097 */     private org.apache.spark.sql.types.StructType aggregateBufferSchema =
/* 098 */     
/* 099 */     new org.apache.spark.sql.types.StructType()
/* 100 */     .add("sum", org.apache.spark.sql.types.DataTypes.LongType)
/* 101 */     .add("sum", org.apache.spark.sql.types.DataTypes.LongType)
/* 102 */     .add("sum", org.apache.spark.sql.types.DataTypes.LongType)
/* 103 */     .add("sum", org.apache.spark.sql.types.DataTypes.DoubleType)
/* 104 */     .add("sum", org.apache.spark.sql.types.DataTypes.DoubleType);
/* 105 */     
/* 106 */     public agg_VectorizedHashMap() {
/* 107 */       batch = org.apache.spark.sql.execution.vectorized.ColumnarBatch.allocate(schema,
/* 108 */         org.apache.spark.memory.MemoryMode.ON_HEAP, capacity);
/* 109 */       // TODO: Possibly generate this projection in TungstenAggregate directly
/* 110 */       aggregateBufferBatch = org.apache.spark.sql.execution.vectorized.ColumnarBatch.allocate(
/* 111 */         aggregateBufferSchema, org.apache.spark.memory.MemoryMode.ON_HEAP, capacity);
/* 112 */       for (int i = 0 ; i < aggregateBufferBatch.numCols(); i++) {
/* 113 */         aggregateBufferBatch.setColumn(i, batch.column(i+5));
/* 114 */       }
/* 115 */       
/* 116 */       buckets = new int[numBuckets];
/* 117 */       java.util.Arrays.fill(buckets, -1);
/* 118 */     }
/* 119 */     
/* 120 */     public org.apache.spark.sql.execution.vectorized.ColumnarBatch.Row findOrInsert(UTF8String agg_key, int agg_key1, double agg_key2, float agg_key3, boolean agg_key4) {
/* 121 */       long h = hash(agg_key, agg_key1, agg_key2, agg_key3, agg_key4);
/* 122 */       int step = 0;
/* 123 */       int idx = (int) h & (numBuckets - 1);
/* 124 */       while (step < maxSteps) {
/* 125 */         // Return bucket index if it's either an empty slot or already contains the key
/* 126 */         if (buckets[idx] == -1) {
/* 127 */           if (numRows < capacity) {
/* 128 */             batch.column(0).putByteArray(numRows, agg_key.getBytes());
/* 129 */             batch.column(1).putInt(numRows, agg_key1);
/* 130 */             batch.column(2).putDouble(numRows, agg_key2);
/* 131 */             batch.column(3).putFloat(numRows, agg_key3);
/* 132 */             batch.column(4).putBoolean(numRows, agg_key4);
/* 133 */             batch.column(5).putNull(numRows);
/* 134 */             batch.column(6).putNull(numRows);
/* 135 */             batch.column(7).putNull(numRows);
/* 136 */             batch.column(8).putNull(numRows);
/* 137 */             batch.column(9).putNull(numRows);
/* 138 */             buckets[idx] = numRows++;
/* 139 */             batch.setNumRows(numRows);
/* 140 */             aggregateBufferBatch.setNumRows(numRows);
/* 141 */             return aggregateBufferBatch.getRow(buckets[idx]);
/* 142 */           } else {
/* 143 */             // No more space
/* 144 */             return null;
/* 145 */           }
/* 146 */         } else if (equals(idx, agg_key, agg_key1, agg_key2, agg_key3, agg_key4)) {
/* 147 */           return aggregateBufferBatch.getRow(buckets[idx]);
/* 148 */         }
/* 149 */         idx = (idx + 1) & (numBuckets - 1);
/* 150 */         step++;
/* 151 */       }
/* 152 */       // Didn't find it
/* 153 */       return null;
/* 154 */     }
/* 155 */     
/* 156 */     private boolean equals(int idx, UTF8String agg_key, int agg_key1, double agg_key2, float agg_key3, boolean agg_key4) {
/* 157 */       return batch.column(0).getUTF8String(buckets[idx]) == agg_key && batch.column(1).getInt(buckets[idx]) == agg_key1 && batch.column(2).getDouble(buckets[idx]) == agg_key2 && batch.column(3).getFloat(buckets[idx]) == agg_key3 && batch.column(4).getBoolean(buckets[idx]) == agg_key4;
/* 158 */     }
/* 159 */     
/* 160 */     private long hash(UTF8String agg_key, int agg_key1, double agg_key2, float agg_key3, boolean agg_key4) {
/* 161 */       long agg_hash = 0;
/* 162 */       
/* 163 */       int agg_result = 0;
/* 164 */       byte[] agg_bytes = agg_key.getBytes();
/* 165 */       for (int i = 0; i < agg_bytes.length; i++) {
/* 166 */         int agg_hash1 = agg_bytes[i];
/* 167 */         agg_result = (agg_result ^ (0x9e3779b9)) + agg_hash1 + (agg_result << 6) + (agg_result >>> 2);
/* 168 */       }
/* 169 */       
/* 170 */       agg_hash = (agg_hash ^ (0x9e3779b9)) + agg_result + (agg_hash << 6) + (agg_hash >>> 2);
/* 171 */       
/* 172 */       int agg_result1 = agg_key1;
/* 173 */       agg_hash = (agg_hash ^ (0x9e3779b9)) + agg_result1 + (agg_hash << 6) + (agg_hash >>> 2);
/* 174 */       
/* 175 */       long agg_result2 = Double.doubleToLongBits(agg_key2);
/* 176 */       agg_hash = (agg_hash ^ (0x9e3779b9)) + agg_result2 + (agg_hash << 6) + (agg_hash >>> 2);
/* 177 */       
/* 178 */       int agg_result3 = Float.floatToIntBits(agg_key3);
/* 179 */       agg_hash = (agg_hash ^ (0x9e3779b9)) + agg_result3 + (agg_hash << 6) + (agg_hash >>> 2);
/* 180 */       
/* 181 */       int agg_result4 = agg_key4 ? 1 : 0;
/* 182 */       agg_hash = (agg_hash ^ (0x9e3779b9)) + agg_result4 + (agg_hash << 6) + (agg_hash >>> 2);
/* 183 */       
/* 184 */       return agg_hash;
/* 185 */     }
/* 186 */     
/* 187 */     public java.util.Iterator<org.apache.spark.sql.execution.vectorized.ColumnarBatch.Row>
/* 188 */     rowIterator() {
/* 189 */       return batch.rowIterator();
/* 190 */     }
/* 191 */     
/* 192 */     public void close() {
/* 193 */       batch.close();
/* 194 */     }
/* 195 */     
/* 196 */   }
```

**Comment by sameeragarwal** [comment]
test this please

**Comment by sameeragarwal** [comment]
cc @nongli @davies

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -267,6 +267,38 @@ class CodegenContext {
...
+   */
+  def setValue(batch: String, row: String, dataType: DataType, ordinal: Int,
+      value: String): String = {
+    val jt = javaType(dataType)
+    dataType match {
+      case _ if isPrimitiveType(jt) =>
+        s"$batch.column($ordinal).put${primitiveTypeName(jt)}($row, $value);"
+      case t: DecimalType => s"$batch.column($ordinal).putDecimal($row, $value, ${t.precision});"
+      case t: StringType => s"$batch.column($ordinal).putByteArray($row, $value.getBytes());"
+      case _ => throw new IllegalArgumentException("cannot generate code for unsupported type")
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Could you put the data type in error message?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 2]
```diff
@@ -267,6 +267,38 @@ class CodegenContext {
...
+    }
+  }
+
+  /**
+   * Returns the specialized code to access a value from a column vector for a given `DataType`.
+   */
+  def getValue(batch: String, row: String, dataType: DataType, ordinal: Int): String = {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Should this just be `getValue(s"$batch.column($ordinal)", dataTpye, row)` ?

**Comment by davies** [comment]
@sameeragarwal Could you add some benchmark for StringType and DecimalType to make sure that we have performance gain for them?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 2 above)
Comment (sameeragarwal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: As discussed offline, let's have a separate method call for clarity (and avoid accidental abuse of `rowIds`/`ordinals`).

**Comment by sameeragarwal** [comment]
Thanks @davies, comments addressed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 3]
```diff
@@ -271,6 +271,74 @@ class CodegenContext {
...
+   */
+  def updateColumn(
+      batch: String,
+      row: String,
+      dataType: DataType,
+      ordinal: Int,
+      ev: ExprCode,
+      nullable: Boolean): String = {
+    if (nullable) {
+      // Can't call setNullAt on DecimalType, because we need to keep the offset
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: For batch, this is not true.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/VectorizedHashMapGenerator.scala [hunk 4]
```diff
@@ -40,12 +41,32 @@ import org.apache.spark.sql.types.StructType
...
     generatedClassName: String,
     groupingKeySchema: StructType,
     bufferSchema: StructType) {
-  val groupingKeys = groupingKeySchema.map(k => (k.dataType.typeName, ctx.freshName("key")))
-  val bufferValues = bufferSchema.map(k => (k.dataType.typeName, ctx.freshName("value")))
-  val groupingKeySignature = groupingKeys.map(_.productIterator.toList.mkString(" ")).mkString(", ")
+  case class Buffer(dataType: DataType, name: String)
+  val groupingKeys = groupingKeySchema.map(k => Buffer(k.dataType, ctx.freshName("key")))
+  val bufferValues = bufferSchema.map(k => Buffer(k.dataType, ctx.freshName("value")))
+  val groupingKeySignature = groupingKeys.map(key => (ctx.javaType(key.dataType), key.name))
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/VectorizedHashMapGenerator.scala]: `s"${ctx.javaType(key.dataType)} ${key.name}"` will be easier to understand

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/VectorizedHashMapGenerator.scala [hunk 5]
```diff
@@ -69,17 +90,29 @@ class VectorizedHashMapGenerator(
     val generatedSchema: String =
       s"""
          |new org.apache.spark.sql.types.StructType()
-         |${(groupingKeySchema ++ bufferSchema).map(key =>
-          s""".add("${key.name}", org.apache.spark.sql.types.DataTypes.${key.dataType})""")
-          .mkString("\n")};
+         |${(groupingKeySchema ++ bufferSchema).map { key =>
+            key.dataType match {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/VectorizedHashMapGenerator.scala]: more `|` ?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/VectorizedHashMapGenerator.scala]: It's to pull this big expression out of the string

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/VectorizedHashMapGenerator.scala [hunk 6]
```diff
@@ -69,17 +90,29 @@ class VectorizedHashMapGenerator(
...
+          }.mkString("\n")};
       """.stripMargin
 
     val generatedAggBufferSchema: String =
       s"""
          |new org.apache.spark.sql.types.StructType()
-         |${bufferSchema.map(key =>
-        s""".add("${key.name}", org.apache.spark.sql.types.DataTypes.${key.dataType})""")
-        .mkString("\n")};
+         |${bufferSchema.map { key =>
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/VectorizedHashMapGenerator.scala]: same this one

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala [hunk 7]
```diff
@@ -224,6 +224,127 @@ class BenchmarkWholeStageCodegen extends SparkFunSuite {
...
+    benchmark.run()
+
+    /*
+    Java HotSpot(TM) 64-Bit Server VM 1.8.0_73-b02 on Mac OS X 10.11.4
+    Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz
+    Aggregate w decimal key:             Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
+    -------------------------------------------------------------------------------------------
+    codegen = F                              6876 / 7216          3.1         327.8       1.0X
+    codegen = T hashmap = F                  5297 / 5478          4.0         252.6       1.3X
+    codegen = T hashmap = T                  5395 / 5668          3.9         257.2       1.3X
```
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala]: Why we does not see improvements here? two many columns or depends on types?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/VectorizedHashMapGenerator.scala [hunk 8]
```diff
@@ -238,4 +281,42 @@ class VectorizedHashMapGenerator(
        |}
      """.stripMargin
   }
+
+  private def computeHash(
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/VectorizedHashMapGenerator.scala]: genComputeHash ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/VectorizedHashMapGenerator.scala [hunk 9]
```diff
@@ -238,4 +281,42 @@ class VectorizedHashMapGenerator(
...
+      input: String,
+      dataType: DataType,
+      result: String,
+      ctx: CodegenContext): String = {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/VectorizedHashMapGenerator.scala]: We usually put `ctx` as the first argument

File: sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala (hunk 7 above)
Comment (sameeragarwal) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/BenchmarkWholeStageCodegen.scala]: I believe it depends on types (string for instance is quite slow as one would expect). With the new changes, the current benchmark is now around 13% faster (152.8 ns/row). Furthermore, if we remove the `StringType`, it actually gets 2.3x faster (68.3 ns/row).

### Example 1:

```scala
sqlContext.range(N)
      .selectExpr(
        "id",
        "(id & 1023) as k1",
        "cast(id & 1023 as string) as k2",  <--------------- [string key]
        "cast(id & 1023 as int) as k3",
        "cast(id & 1023 as double) as k4",
        "cast(id & 1023 as float) as k5",
        "id > 1023 as k6")
      .groupBy("k1", "k2", "k3", "k4", "k5", "k6")
      .sum()
      .collect()
```

    Java HotSpot(TM) 64-Bit Server VM 1.8.0_73-b02 on Mac OS X 10.11.4
    Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz
    Aggregate w decimal key:             Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    -------------------------------------------------------------------------------------------
    codegen = F                              5885 / 6091          3.6         280.6       1.0X
    codegen = T hashmap = F                  3625 / 4009          5.8         172.8       1.6X
    codegen = T hashmap = T                  3204 / 3271          6.5         152.8       1.8X

### Example 2:

```scala
  sqlContext.range(N)
      .selectExpr(
        "id",
        "(id & 1023) as k1",
        "cast(id & 1023 as long) as k2",  <--------------- [long key]
        "cast(id & 1023 as int) as k3",
        "cast(id & 1023 as double) as k4",
        "cast(id & 1023 as float) as k5",
        "id > 1023 as k6")
      .groupBy("k1", "k2", "k3", "k4", "k5", "k6")
      .sum()
      .collect()
```

    Java HotSpot(TM) 64-Bit Server VM 1.8.0_73-b02 on Mac OS X 10.11.4
    Intel(R) Core(TM) i7-4960HQ CPU @ 2.60GHz
    Aggregate w multiple keys:          Best/Avg Time(ms)    Rate(M/s)   Per Row(ns)   Relative
    -------------------------------------------------------------------------------------------
    codegen = F                              3090 / 3687          6.8         147.3       1.0X
    codegen = T hashmap = F                  2028 / 2142         10.3          96.7       1.5X
    codegen = T hashmap = T                  1433 / 1543         14.6          68.3       2.2X

**Comment by davies** [comment]
LGTM, 
Merging this into master, thanks!

**Pull request closed without merging**